hpedockerplugin/request_validator.py

```diff
@@ -21,116 +8,147 @@
-                  'mountConflictDelay']
-    _validate_opts("create snapshot", contents, valid_opts)
-
-
-def _validate_snapshot_schedule_opts(contents):
-    valid_opts = ['virtualCopyOf', 'scheduleFrequency', 'scheduleName',
-                  'snapshotPrefix', 'expHrs', 'retHrs', 'mountConflictDelay']
-    mandatory_opts = ['scheduleName', 'snapshotPrefix', 'scheduleFrequency']
-    _validate_opts("create snapshot schedule", contents,
-                   valid_opts, mandatory_opts)
-
-
-def _validate_import_vol_opts(contents):
-    valid_opts = ['importVol', 'backend', 'mountConflictDelay']
-    _validate_opts("import volume", contents, valid_opts)
+    def _validate_create_volume_opts(self, contents):
+        valid_opts = ['compression', 'size', 'provisioning',
+                      'flash-cache', 'qos-name', 'fsOwner',
+                      'fsMode', 'mountConflictDelay', 'cpg',
+                      'snapcpg', 'backend']
+        self._validate_opts("create volume", contents, valid_opts)
+
+    def _validate_clone_opts(self, contents):
+        valid_opts = ['cloneOf', 'size', 'cpg', 'snapcpg',
+                      'mountConflictDelay']
+        self._validate_opts("clone volume", contents, valid_opts)
+
+    def _validate_snapshot_opts(self, contents):
+        valid_opts = ['virtualCopyOf', 'retentionHours', 'expirationHours',
+                      'mountConflictDelay']
+        self._validate_opts("create snapshot", contents, valid_opts)
+
+    def _validate_snapshot_schedule_opts(self, contents):
+        valid_opts = ['virtualCopyOf', 'scheduleFrequency', 'scheduleName',
+                      'snapshotPrefix', 'expHrs', 'retHrs',
+                      'mountConflictDelay']
+        mandatory_opts = ['scheduleName', 'snapshotPrefix',
+                          'scheduleFrequency']
+        self._validate_opts("create snapshot schedule", contents,
+                            valid_opts, mandatory_opts)
+
+    def _validate_import_vol_opts(self, contents):
+        valid_opts = ['importVol', 'backend', 'mountConflictDelay']
+        self._validate_opts("import volume", contents, valid_opts)
+
+        # Replication enabled backend cannot be used for volume import
+        if 'Opts' in contents:
+            backend_name = contents['Opts'].get('backend', None)
+            if not backend_name:
+                backend_name = 'DEFAULT'
+            try:
+                config = self._backend_configs[backend_name]
+            except KeyError:
+                backend_names = list(self._backend_configs.keys())
+                backend_names.sort()
+                msg = "ERROR: Backend '%s' doesn't exist. Available " \
+                      "backends are %s. Please use " \
+                      "a valid backend name and retry." % \
+                      (backend_name, backend_names)
+                raise exception.InvalidInput(reason=msg)
 
+            if config.replication_device:
```

Comment:
@imran-ansari @wdurairaj I think we should not completely block the importing of a volume if a configuration has replication_device defined in its backend.
This should only be blocked if imported volume also needs to be replicated.

The current introduced check will block all volumes from import if a backend is configured for replication device.
Ex
The user creates a volume v1 and v2 on backend b1 (has replication_device), but v2 needs to be replicated and v1 is normal volume. (This is a valid use case).
Now if there is a 3PAR volume v3 which needs to be imported to this backend, the plugin should allow this to happen as long as there is no replicationGroup specified.

Reply:
Normal volumes cannot be created using replicated backend. For that, user must create another backend which is not replication enabled.

Reply:
@imran-ansari I missed the part where we already have a check to not create volumes when a replicationGroup is not specified but the backend is configured for replication. I am okay with the current change.

Reply:
Thanks Swapnil!